File: sources/scala/13_matrix_mul/main.scala
// task 13 matrix_mul — expected output: 599995000
// build: scalac -release 17 main.scala    run: scala Main

object Main {
  def main(args: Array[String]): Unit = {
    val n = 500
    val size = n * n
    val a = new Array[Long](size)
    val b = new Array[Long](size)
    val c = new Array[Long](size)

    var i = 0
    while (i < n) {
      var j = 0
      while (j < n) {
        a(i * n + j) = ((i + j) % 7).toLong
        j += 1
      }
      i += 1
    }

    i = 0
    while (i < n) {
      var j = 0
      while (j < n) {
        b(i * n + j) = ((i * j) % 5).toLong
        j += 1
      }
      i += 1
    }

    i = 0
    while (i < n) {
      var j = 0
      while (j < n) {
        var sum = 0L
        var k = 0
        while (k < n) {
          sum += a(i * n + k) * b(k * n + j)
          k += 1
        }
        c(i * n + j) = sum
        j += 1
      }
      i += 1
    }

    var total = 0L
    i = 0
    while (i < size) {
      total += c(i)
      i += 1
    }
    println(total)
  }
}

File: sources/kotlin/13_matrix_mul/main.kt
// task 13 matrix_mul — expected output: 599995000
// build: kotlinc main.kt -include-runtime -d prog.jar    run: java -jar prog.jar    [native build: kotlinc-native -opt -o prog main.kt    native run: ./prog]

fun main() {
    val n = 500
    val a = LongArray(n * n)
    val b = LongArray(n * n)
    val c = LongArray(n * n)
    for (i in 0 until n) {
        for (j in 0 until n) {
            a[i * n + j] = ((i + j) % 7).toLong()
            b[i * n + j] = ((i * j) % 5).toLong()
        }
    }
    for (i in 0 until n) {
        for (j in 0 until n) {
            var sum = 0L
            for (k in 0 until n) {
                sum += a[i * n + k] * b[k * n + j]
            }
            c[i * n + j] = sum
        }
    }
    var total = 0L
    for (i in 0 until n * n) {
        total += c[i]
    }
    println(total)
}
